File: LeetCode Daily/00725. Split Linked List in Parts.ts
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     val: number
 *     next: ListNode | null
 *     constructor(val?: number, next?: ListNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.next = (next===undefined ? null : next)
 *     }
 * }
 */

function splitListToParts(
    head: ListNode | null,
    k: number,
): Array<ListNode | null> {
    let length = 0;
    let current = head;
    while (current) {
        length++;
        current = current.next;
    }

    let result = new Array(k).fill(null);
    let partLength = Math.floor(length / k);
    let extra = length % k;

    current = head;
    for (let i = 0; i < k; i++) {
        if (!current) break;
        result[i] = current;
        let partSize = partLength + (extra-- > 0 ? 1 : 0);
        for (let j = 0; j < partSize - 1; j++) {
            current = current.next;
        }
        let next = current.next;
        current.next = null;
        current = next;
    }

    return result;
}

File: LeetCode Daily/00725. Split Linked List in Parts.cs
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     public int val;
 *     public ListNode next;
 *     public ListNode(int val=0, ListNode next=null) {
 *         this.val = val;
 *         this.next = next;
 *     }
 * }
 */
public class Solution
{
    public class ListNode
    {
        public int val;
        public ListNode next;
        public ListNode(int val = 0, ListNode next = null)
        {
            this.val = val;
            this.next = next;
        }
    }

    public ListNode[] SplitListToParts(ListNode head, int k)
    {
        var result = new ListNode[k];
        var length = 0;
        var current = head;
        while (current != null)
        {
            length++;
            current = current.next;
        }

        var size = length / k;
        var remainder = length % k;
        current = head;
        for (var i = 0; i < k; i++)
        {
            var count = size + (remainder > 0 ? 1 : 0);
            remainder--;
            var dummy = new ListNode();
            var node = dummy;
            for (var j = 0; j < count; j++)
            {
                node.next = new ListNode(current.val);
                node = node.next;
                current = current.next;
            }

            result[i] = dummy.next;
        }

        return result;
    }
}